pants

```diff
@@ -73,23 +73,61 @@ EOF
 
 # The high-level flow:
 # 1.) Resolve the Python interpreter, first reading from the env var $PYTHON,
-#     then reading from pants.ini, then defaulting to `python2.7`.
+#     then reading from pants.ini, then determining the default based off of
+#     which Python versions the current `pants_version` supports.
 # 2.) Grab pants version from pants.ini or default to latest.
 # 3.) Check for a venv via a naming/path convention and execute if found.
 # 4.) Otherwise create venv and re-exec self.
 #
 # After that pants itself will handle making sure any requested plugins
 # are installed and up to date.
 
-function determine_pants_runtime_python_version {
+function determine_default_python_exe {
+  pants_version="$(get_pants_ini_config_value 'pants_version')"
+  # If `pants_version` not pinned in `pants.ini`, default to what the most
+  # recent stable release supports.
+  if [[ -z "${pants_version}" ]]; then
+    supported_python_versions=('3.6' '2.7')
+    supported_message='2.7 or 3.6'
+  else
+    pants_major_version="$(echo "${pants_version}" | cut -d '.' -f1)"
+    pants_minor_version="$(echo "${pants_version}" | cut -d '.' -f2)"
+    if [[ "${pants_major_version}" -eq 2 ]]; then
+      supported_python_versions=('3.6' '3.7' '3.8' '3.9')
```

Comment:
@stuhood should we make any assumptions about if we will support Python 3.6 in the 2.0 release? 

The main things we'd get by dropping Python 3.6 are context variables, `from __future__ import annotations` (useful for type hints) and improved `asyncio`. Python 3.6 is EOL in December 2021. I don't image we'd drop it until 2021 at the earliest for that reason.

Reply:
I don't think we should make any assumptions. Probably easiest to just have "everything greater than known version X" default to the max values for version X (ie `1.17.0` in this case).